### PR TITLE
Add validator to return a warning when `state-into-spec` is set to `merge`

### PR DESCRIPTION
**Pull request by maqiuyujoyce**
### Change description

Added a validating webhook that checks whether the `state-into-spec` annotation is explicitly set to `merge` and returns a warning if so.

Further improvement would be validating whether the `state-into-spec` annotation will be defaulted to `merge` for a resource, and/or validating whether `stateIntoSpec` field is set to `Merge` for CC or CCC.

### Tests you have done

- [N/A] Run `make ready-pr` to ensure this PR is ready for review.
- [X] Perform necessary E2E testing for changed resources.

Ran `make docker-build docker-push deploy` to install a locally built version, and verified that the warning is returned when `cnrm.cloud.google.com/state-into-spec: merge` is set:

```
$ kubectl apply -f pubsubtopic-merge.yaml 
Warning: 'cnrm.cloud.google.com/state-into-spec: merge' is unsupported for CRDs added in 1.114.0 and later. Use 'absent' instead. More details can be found at https://cloud.google.com/config-connector/docs/concepts/ignore-unspecified-fields.
pubsubtopic.pubsub.cnrm.cloud.google.com/pubsubtopic-sample-2 created
```

### CODE REVIEW

File: pkg/webhook/state_into_spec_annotation_validator.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type stateIntoSpecAnnotationValidator struct {
+	client client.Client
+}
+
+func NewStateIntoSpecAnnotationValidatorHandler() HandlerFunc {
```
Comment (yuwenma) [inline pkg/webhook/state_into_spec_annotation_validator.go:35]: nit: docstring for public methods.

File: pkg/webhook/state_into_spec_annotation_validator.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		klog.Error(err)
+		return admission.Errored(http.StatusBadRequest,
+			fmt.Errorf("error decoding object: %w", err))
+	}
+
+	value, ok := k8s.GetAnnotation(k8s.StateIntoSpecAnnotation, obj)
+	if ok && value == k8s.StateMergeIntoSpec {
+		return allowedResponse.WithWarnings(
+			fmt.Sprintf("'%v: %v' is unsupported for CRDs added in "+
+				"1.114.0 and later. Use '%v' instead. More details can be "+
```
Comment (yuwenma) [inline pkg/webhook/state_into_spec_annotation_validator.go:54]: Shall we suggest not using the `state-into-spec` annotation instead? (compared to using "absent" value)

Comment (maqiuyujoyce) [reply pkg/webhook/state_into_spec_annotation_validator.go:54]: At the moment, not using `state-into-spec` annotation means let KCC defaults the value, which is still `merge` for most resources. So I added a link to the doc which suggests users default it to `Absent` at cluster/namespace-level - this is the most recommended behavior right now.

File: pkg/webhook/state_into_spec_annotation_validator.go (hunk 1 above)
Comment (maqiuyujoyce) [reply pkg/webhook/state_into_spec_annotation_validator.go:38]: Done.

**Comment by gemmahou** [comment]
/lgtm
/approve

Looks good to me!

**Pull request merged**